src/cmdline/mod.rs

```diff
@@ -345,42 +290,216 @@ impl Cmdline {
+    /// https://elixir.bootlin.com/linux/v5.10.139/source/kernel/params.c#L179
+    ///
+    /// To get bytes of the cmdline to be written in guest's memory (including the
+    /// null terminator) from this representation, use CString::as_bytes_with_nul()
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use linux_loader::cmdline::*;
+    /// let mut cl = Cmdline::new();
+    /// cl.insert_string("foobar");
+    /// assert_eq!(cl.as_cstring().unwrap().as_bytes_with_nul(), b"foobar\0");
+    /// ```
+    pub fn as_cstring(&self) -> Result<CString> {
+        if self.boot_args.0.is_empty() {
+            return if self.init_args.0.is_empty() {
+                CString::new("".to_string()).map_err(|_| Error::NullTerminator)
+            } else {
+                Err(Error::NoBootArgsInserted)
+            }
+        }
+
+        if self.init_args.0.is_empty() {
+            if self.boot_args.0.len() + 1 > CMDLINE_MAXSIZE {
+                // need space for the null terminator
+                Err(Error::TooLarge)
+            } else {
+                CString::new(self.boot_args.0.to_string()).map_err(|_| Error::NullTerminator)
+            }
+        } else if self.boot_args.0.len() + self.init_args.0.len() + 5 > CMDLINE_MAXSIZE {
+            // need space for the null terminator and ' -- ' delimiter
+            Err(Error::TooLarge)
+        } else {
+            CString::new(format!("{} -- {}", self.boot_args.0, self.init_args.0)).map_err(|_| Error::NullTerminator)
+        }
+    }
+}
+
+impl TryFrom<String> for Cmdline {
+    type Error = String;
+
+    fn try_from(cmdline_raw: String) -> result::Result<Self, Self::Error> {
+        // Checking for ' -- ' sequence is required as long as a sequence like 'foo-- bar' should not be
+        // handled as a boot param 'foo' and an init arg 'bar'
+        let (mut boot_args, mut init_args) = match cmdline_raw.split_once(" -- ") {
+            None => {
+                // We still need to check is the given cmdline isn't just build from the init args part
+                match cmdline_raw.split_once("-- ") {
+                    None => (cmdline_raw.as_str(), ""),
+                    Some((boot_args_section, init_args_section)) => {
+                        // We have to check that '-- ' is at the beginning of cmdline, otherwise it is part
+                        // of a boot argument and should NOT be handled as delimiter
+                        if boot_args_section.is_empty() {
+                            ("", init_args_section)
+                        } else {
+                            (cmdline_raw.as_str(), "")
+                        }
+                    }
+                }
+            }
+            Some((boot_args_section, init_args_section)) => (boot_args_section, init_args_section),
+        };
+
+        boot_args = boot_args.trim();
+        init_args = init_args.trim();
+
+        let mut more_space = 1; // required for null terminator of the cmdline
```

Comment:
It would make the code more readable if we instead define a `cmdline_size` instead of `more_space`. Then, the code will read more like a story:
- cmdline size initially is boot_args.len() + 1;
- if in the cmdline we also have `init_args`, then, the cmdline_size += init_args.len() + INIT_ARGS_DELIMITER.len().
- if the cmdline size is bigger than CMDLINE_MAXSIZE, then we return an error

Reply:
updating that